Added instructions for sdk local workflow (#11868)
Azure/azure-sdk-tools: eng/common/instructions/azsdk-tools/local-sdk-workflow.instructions.md

---

> description: "Guide the user to generate and build SDKs locally for TypeSpec changes"

Reviewer: ```suggestion
description: "Guide the user to generate and build SDKs locally for a TypeSpec based API spec"
```

---

# Goal

> Help the user generate and build SDKs locally from TypeSpec API specifications using the `azure-sdk-mcp` tools.

Reviewer 2: May I know how customers would refer these instructions? I may not quite understand the context here, are these instructions used in customers' IDEs?

---

## Part A: Generate SDK Locally

### Step 1: Outline workflow

**Goal**: Ensure the user understands the overall SDK generation and build process before starting.  
**Actions**:

- Present the high-level steps involved in generating and building SDK locally:
  1. Select target language
  2. Verify SDK repository
  3. Validate repository path
  4. Identify path to configuration file

>   5. Generate SDK using `azure-sdk-mcp`

Reviewer 3: what is `azure-sdk-mcp`? Are you referring the mcp tool name in the mcp.json? Can we avoid that to remove that dependency with the configuration name.

Author: Good reminder. I updated it with the name of MCP tools.

  6. Identify SDK project path
  7. Build/Compile SDK using `azure-sdk-mcp`
- Ask the user to confirm readiness to proceed.

---

### Step 2: Select language

**Goal**: Confirm the target language for SDK generation.  
**Actions**:

- Prompt user to choose one of the supported languages:
  - .NET
  - Java
  - JavaScript
  - Python
  - Go
- Validate input against the allowed list.

---

### Step 3: Verify SDK repository

**Goal**: Ensure the correct Azure SDK language repository is available locally.
**Actions**:

- Prompt user to provide the path to their **locally cloned repository** for the selected language.
- Note: The **local folder name can be arbitrary**, but the repository must have originated from one of the official Azure SDK repositories:
  - `azure-sdk-for-net` (.NET)
  - `azure-sdk-for-java` (Java)
  - `azure-sdk-for-js` (JavaScript)
  - `azure-sdk-for-python` (Python)
  - `azure-sdk-for-go` (Go)
- If the repository is not cloned → instruct user to clone the appropriate remote repository from GitHub.
- MCP tool will automatically validate the remote origin and repository structure.

---

### Step 4: Validate repository path

**Actions**:

> - Check if the provided repository path exists and matches the selected SDK language repository.

Reviewer 3: Do we need to verify the setup in language repo? I remember @LibbaLawrence had a similar step in old version of local SDK generation for Python

Author: Setup verification is not covered here. When we have the `setup-env` tool ready, it will be chained to these two tools if the error is related to env issue.

> Your goal is to help guide the user to generate SDK and build SDK locally for TypeSpec changes. This is currently supported for five languages: .NET, Java, JavaScript(TypeScript), Python, and Go.

Reviewer: does it need to know that this is only for TypeSpec changes?
If so, maybe we should say the API spec is TypeSpec based? but also, not sure how the agent will know ...

Author: It doesn't have any difference `w/ or w/t` 'TypeSpec' word here since the `copilot-instructions.md` already has this constraint: `Steps to generate SDK from TypeSpec API specification`. If I explicitly asked agent to generate sdk for swagger spec, it would recommend users to convert swagger to typespec.
<img width="927" height="825" alt="image" src="https://github.com/user-attachments/assets/328a8733-3d24-41aa-a58e-07a086897f75" />

> Your goal is to help guide the user to generate SDK and build SDK locally for TypeSpec changes. This is currently supported for five languages: .NET, Java, JavaScript(TypeScript), Python, and Go.

Author: Each language repo will need to reference this file in its 'copilot-iinstructions.md'. Similar to this:
```
## Local SDK Generation and Build Guidelines
**ALWAYS use Azure SDK MCP tools first** when available for SDK-related tasks: generate code, build code, validation.
ACTION: follow [local-sdk-workflow](..\eng\common\instructions\azsdk-tools\local-sdk-workflow.instructions.md).
```

> ## Steps to generate SDK locally from TypeSpec

Reviewer: ```suggestion
## Steps to generate SDK locally from a TypeSpec API specification
```

### Step 1: Tell user the steps to generate SDK locally

> - specify the language for SDK generation

Reviewer: ```suggestion
- Identify the language for SDK generation
```

> - have the corresponding azure-sdk-for-{language} repository cloned locally

Reviewer: ```suggestion
- Validate the corresponding azure-sdk-for-{language} repository is present in local environment
```

- specify the path to the cloned azure-sdk-for-{language} repository
- specify the path to the 'tspconfig.yaml' file or the path to 'tsp-location.yaml' file.
- use tools provided by 'azure-sdk-mcp' MCP server to generate SDK locally

### Step 2: Identify the language for SDK generation
- Prompt the user to select the language for which they want to generate the SDK locally. Supported languages are:
    - .NET
    - Java
    - JavaScript (TypeScript)
    - Python
    - Go

### Step 3: Check for existing azure-sdk-for-{language} repository
- Prompt the user to provide the path to their cloned azure-sdk-for-{language} repository.
- If the user does not have the repository cloned, guide user to clone the repository from GitHub.

### Step 4: Validate repository path

> - Check if the provided path is valid. If not, prompt the user to provide a valid path.

Reviewer: ```suggestion
- Check if the provided repository path is valid. If not, prompt the user to provide a valid path for the repository's location.
```

### Step 5: Prepare other inputs and validate

> - If user is within the azure-rest-api-specs repository, skip 'tsp-location.yaml' input and use the path to 'tspconfig.yaml' file.

Reviewer: but if the user is in this context, then how are the SDKs going to be generated locally? shouldn't it be the other way where if the user is on the specs repo the flow should ask the user to move to the cloned repository?

- If user is within the azure-sdk-for-{language} repository, skip 'tspconfig.yaml' input and use the path to 'tsp-location.yaml' file.

> - The path to 'tspconfig.yaml' can be a local path or a HTTPS URL.

Reviewer: I got lost here :( where is the user giving the URL for the tspconfig.yaml?

Author: No matter which folder the tool is run from, users should be able to ask the agent to _generate sdk for me for this spec(which is a remote URL)_. The tool will generate code from the remote spec as long as users have the corresponding SDK repo cloned locally.
This is the experience we want to support for SDK generation based on a remote 'tspconfig.yaml'.  CC: @lmazuel

> - When the provided tspConfigPath is a local path, the other two inputs must be provided:

Reviewer: there is no reference to what `tspConfigPath ` is. Is this an implementation detail of the MCP?

    - Full name of the repository in 'owner/repo' format
    - Commit SHA of the 'azure-rest-api-specs' repository

> ### Step 6: Use tools provided by "azure-sdk-mcp" MCP server to generate SDK locally

Reviewer: should we point to the starting point tool? or will the agent know

Author: Agent can detect current MCP tools installed. I don't introduce the dependency on the tool's name here.

>     - Commit SHA of the 'azure-rest-api-specs' repository

Author: When the provided tspconfig path is a local file path, it will generate code based on the current specs at the local repo path. 

The commit is used to fill in to 'tsp-location.yaml'. An idea is to leave the 'commit' property as empty in 'tsp-location.yaml' then user will have to update it when work on the SDK PR.

>     - Commit SHA of the 'azure-rest-api-specs' repository

Author: I removed them from the input.

> ## Steps to build SDK locally

Reviewer: is this a flow for when a suer just wants to build the SDK and work on the SDK right?
It is not a generation, just build

Author: That's correct.
